### ISSUE TITLE
torch.exp() cannot be modified by an inplace operation

### ISSUE THREAD

**Issue by chwan-rice** [pytorch/pytorch#41926]
## 📚 Documentation

The original question was posted [here](https://discuss.pytorch.org/t/torch-exp-is-modified-by-an-inplace-operation/90216). As `torch.exp()` leverages its output for calculating its gradient while most other functions use their input, `torch.exp()` has a different behavior when its output is inplaced, which is not natural to users. I understand that this implementation is beneficial to the code efficiency. The description of `torch.exp()` in the document should emphasize this property.

**Comment by mruberry** [comment]
This is an interesting issue where the user didn't expect that a tensor (the output of exp) was needed for a gradient computation. I wonder what a more general fix for this issue would be? Maybe we just continue to expect users to understand, possibly through trial-and-error, which tensors are needed for the gradient computations they're doing?

**Comment by ezyang** [comment]
Well, the point of version counter checks is that we can report to the user when such a situation happens. We could conceivably also augment documentation to let people know when outputs are saved for backwards, but VC check is our most important line of defense.

**Comment by albanD** [comment]
I do agree that this is an interesting issue as in general, it is an arbitrary implementation choice for such op to either save the input or the output.
In general, we make the choice of saving whatever is needed to make the backward faster to compute.
In this particular case, if we were saving the input, we would need to re-compute the exp() again during the backward leading to double computation.

I do agree that we could do a better job at sharing with the user what gets saved but we would need to extract that information from the codegen logic as we don't have it readily available right now.
